core/queryalgebra/evaluation/src/main/java/org/eclipse/rdf4j/query/algebra/evaluation/impl/ProjectionRemovealOptimizer.java

```diff
@@ -0,0 +1,75 @@
+/*******************************************************************************
+ * Copyright (c) 2019 Eclipse RDF4J contributors.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Distribution License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/org/documents/edl-v10.php.
+ *******************************************************************************/
+package org.eclipse.rdf4j.query.algebra.evaluation.impl;
+
+import java.util.HashSet;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import org.eclipse.rdf4j.query.BindingSet;
+import org.eclipse.rdf4j.query.Dataset;
+import org.eclipse.rdf4j.query.algebra.Projection;
+import org.eclipse.rdf4j.query.algebra.TupleExpr;
+import org.eclipse.rdf4j.query.algebra.Var;
+import org.eclipse.rdf4j.query.algebra.evaluation.QueryOptimizer;
+import org.eclipse.rdf4j.query.algebra.helpers.AbstractQueryModelVisitor;
+
+/**
+ * If a projection node in the algebra does not contribute or change the results it can be removed from the tree.
+ *
+ * For example <code>
+ * SELECT ?s ?p ?o
+ * WHERE {?s ?p ?o }
+ * </code> Does not need a projection as the inner statement pattern returns the same result.
```

Comment:
I haven't looked in detail but I'd like to check specifically how this works in the case of subselects - I seem to remember that we have several optimizers that look for subselects by checking if there is a (nested) projection. We'll have to make sure that those cases continue to work.

Reply:
I think there is an other option but that would require merging #3354 . Then we can mark an algebra step as not needed and remove it from the evaluation.

Reply:
Oh I'm not against this approach per se - I'd just like to be sure it doesn't break subselect-queries. My remark was from the top of my head so I may be way off base.